src/coreComponents/mesh/generators/VTKMeshGenerator.cpp

```diff
@@ -323,270 +304,210 @@ std::vector< T > gatherData( std::vector< T > const & data )
+ElementType convertVtkToGeosxElementType( VTKCellType const cellType )
 {
-  std::map< int, std::vector< vtkIdType > > surfacesIdsToCellsIds;
-
-  vtkIntArray const * const attributeDataArray = getDataArrayOptional< vtkIntArray >( mesh->GetCellData(), "attribute" );
-
-  auto countCell = [&]( localIndex c,
-                        std::map< int, std::vector< vtkIdType > > & regionsElem )
+  switch( cellType )
   {
-    if( attributeDataArray != nullptr )
-    {
-      int const region = attributeDataArray->GetValue( c );
-      GEOSX_LOG_RANK_0_IF( region < 0, "Attribute value " << region << " not supported, please use value >=0 to describe regions" );
-      regionsElem[region].push_back( c );
-    }
-    else
+    case VTK_VERTEX:     return ElementType::Vertex;
+    case VTK_LINE:       return ElementType::Line;
+    case VTK_TRIANGLE:   return ElementType::Triangle;
+    case VTK_QUAD:       return ElementType::Quadrilateral;
+    case VTK_POLYGON:    return ElementType::Polygon;
+    case VTK_TETRA:      return ElementType::Tetrahedron;
+    case VTK_PYRAMID:    return ElementType::Pyramid;
+    case VTK_WEDGE:      return ElementType::Prism;
+    case VTK_HEXAHEDRON: return ElementType::Hexahedron;
+    case VTK_POLYHEDRON: return ElementType::Polyhedron;
+    default:
     {
-      regionsElem[-1].push_back( c );
+      GEOSX_ERROR( cellType << " is not a recognized cell type to be used with the VTKMeshGenerator" );
     }
-  };
-
+  }
+  return ElementType::Polyhedron;
+}
 
-  for( vtkIdType c = 0; c < mesh->GetNumberOfCells(); c++ )
+std::map< ElementType, std::vector< vtkIdType > >
+splitElementsByType( vtkUnstructuredGrid & mesh )
+{
+  std::map< ElementType, std::vector< vtkIdType > > elemsByType;
   {
-    if( VTK_HEXAHEDRON == mesh->GetCellType( c ) )
-    {
-      countCell( c, regionsHex );
-    }
-    else if( VTK_TETRA == mesh->GetCellType( c ) )
+    // Use a temporary array for speed (avoid accessing maps in a hot loop).
+    std::array< std::vector< vtkIdType >, numElementTypes() > cellsByTypeTemp;
+    for( vtkIdType c = 0; c < mesh.GetNumberOfCells(); c++ )
     {
-      countCell( c, regionsTetra );
+      ElementType const elemType = convertVtkToGeosxElementType( static_cast< VTKCellType >( mesh.GetCellType( c ) ) );
+      cellsByTypeTemp[static_cast< integer >( elemType )].push_back( c );
     }
-    else if( VTK_WEDGE == mesh->GetCellType( c ) )
+    // Convert from array to map while performing some checks
+    for( integer t = 0; t < numElementTypes(); ++t )
     {
-      countCell( c, regionsWedges );
-    }
-    else if( VTK_PYRAMID == mesh->GetCellType( c ) )
-    {
-      countCell( c, regionsPyramids );
-    }
-    else if( VTK_TRIANGLE == mesh->GetCellType( c ) || VTK_QUAD == mesh->GetCellType( c ) )
-    {
-      if( attributeDataArray != nullptr )
+      if( cellsByTypeTemp[t].empty() )
```

Comment:
Unfortunately, the code below creates extra entries with empty cell lists in the map, unless I check and continue here. Those entries are mostly harmless, but they will show up for example when I print element type statistics (you'll see something like `Polygon: 0` when no polygons were imported). Instead of adding code to remove "dead" entries later, it's easier to not create them in the first place.